apps/src/lib/node/ledger/shell/prepare_proposal.rs

```diff
@@ -143,58 +168,186 @@ where
             .get_protocol_key()
             .expect("Validators should always have a protocol key");
 
-        iter_protocol_txs(VoteExtensionDigest {
+        let txs: Vec<_> = iter_protocol_txs(VoteExtensionDigest {
             ethereum_events,
             validator_set_update,
         })
         .map(|tx| tx.sign(protocol_key).to_bytes())
-        .collect()
+        .collect();
+
+        // TODO(feature = "abcipp"):
+        // - alloc space for each protocol tx
+        // - handle space allocation errors
+        // - transition to new allocator state
+
+        todo!()
     }
 
     /// Builds a batch of vote extension transactions, comprised of Ethereum
     /// events and, optionally, a validator set update
     #[cfg(not(feature = "abcipp"))]
-    fn build_vote_extension_txs(&mut self, txs: &[TxBytes]) -> Vec<TxBytes> {
-        if self.storage.last_height != BlockHeight(0) {
-            deserialize_vote_extensions(txs).collect()
-        } else {
+    fn build_vote_extension_txs(
+        &mut self,
+        mut alloc: BlockSpaceAllocator<BuildingProtocolTxBatch>,
+        tx_indices: &mut VecIndexSet<u128>,
+        txs: &[TxBytes],
+    ) -> (Vec<TxBytes>, EncryptedTxBatchAllocator) {
+        if self.storage.last_height == BlockHeight(0) {
             // genesis should not contain vote extensions
-            vec![]
+            return (vec![], self.get_encrypted_txs_allocator(alloc));
         }
+
+        let txs = deserialize_vote_extensions(txs, tx_indices).take_while(|tx_bytes|
+            alloc.try_alloc(&*tx_bytes)
+                .map_or_else(
+                    |status| match status {
+                        AllocFailure::Rejected { bin_space_left } => {
+                            // TODO: maybe we should find a way to include
```

Comment:
We could also add prioritization to protocol txs

Reply:
yeah that's what I was thinking of doing, at the `CheckTx` level, but that's for another PR